pkg/storage/gcs.go

```diff
@@ -258,3 +298,82 @@ func hasSSTFiles(ctx context.Context, bucket *storage.BucketHandle, prefix strin
 	}
 	return false
 }
+
+// gcsObjectReader wrap storage.Reader and add the `Seek` method.
+type gcsObjectReader struct {
+	storage   *gcsStorage
+	name      string
+	objHandle *storage.ObjectHandle
+	reader    io.ReadCloser
+	pos       int64
+	// reader context used for implement `io.Seek`
+	// currently, lightning depends on package `xitongsys/parquet-go` to read parquet file and it needs `io.Seeker`
+	// See: https://github.com/xitongsys/parquet-go/blob/207a3cee75900b2b95213627409b7bac0f190bb3/source/source.go#L9-L10
+	ctx context.Context
+}
+
+// Read implement the io.Reader interface.
+func (r *gcsObjectReader) Read(p []byte) (n int, err error) {
+	if r.reader == nil {
+		rc, err := r.objHandle.NewRangeReader(r.ctx, r.pos, -1)
+		if err != nil {
+			return 0, errors.Annotatef(err,
+				"failed to read gcs file, file info: input.bucket='%s', input.key='%s'",
+				r.storage.gcs.Bucket, r.name)
+		}
+		r.reader = rc
+	}
+	n, err = r.reader.Read(p)
+	r.pos += int64(n)
+	return n, err
+}
+
+// Close implement the io.Closer interface.
+func (r *gcsObjectReader) Close() error {
+	if r.reader == nil {
+		return nil
+	}
+	return r.reader.Close()
+}
+
+// Seek implement the io.Seeker interface.
+//
+// Currently, tidb-lightning depends on this method to read parquet file for gcs storage.
+func (r *gcsObjectReader) Seek(offset int64, whence int) (int64, error) {
+	var realOffset int64
+	switch whence {
+	case io.SeekStart:
+		if r.pos < 0 {
+			return 0, errors.Annotatef(berrors.ErrInvalidArgument, "Seek: offset '%v' out of range.", offset)
+		}
+		realOffset = offset
+	case io.SeekCurrent:
+		realOffset = r.pos + offset
+		if r.pos < 0 && realOffset >= 0 {
+			return 0, errors.Annotatef(berrors.ErrInvalidArgument, "Seek: offset '%v' out of range. current pos is '%v'.", offset, r.pos)
+		}
+	case io.SeekEnd:
+		if offset >= 0 {
+			return 0, errors.Annotatef(berrors.ErrInvalidArgument, "Seek: offset '%v' should be negative.", offset)
+		}
+		realOffset = offset
```

Comment:
?

Reply:
see `Example(LastNBytes)` in https://pkg.go.dev/cloud.google.com/go/storage#ObjectHandle.NewRangeReader
this package support negative offset

Reply:
@Leavrth please include the comment https://github.com/pingcap/br/pull/1353/files#r670333536 in the code, as this behavior is quite unintuitive.

Reply:
done.